File: src/TrayAppRUI/Views/MainWindow.axaml.fs
namespace FSharpReactiveUI.Views

open System
open Avalonia.Controls
open Avalonia.Controls.Notifications
open Avalonia.Markup.Xaml

type MainWindow() as this =
    inherit Window()

    let notificationManager = WindowNotificationManager(this, MaxItems = 3)
    let pending = System.Collections.Generic.Queue<Notification>()
    let mutable initialized = false

    do this.InitializeComponent()

    member private _.InitializeComponent() = AvaloniaXamlLoader.Load(this)

    member _.Notify(title, body, ?tpe) =
        let t = defaultArg tpe NotificationType.Information
        let n = Notification(title, body, t, TimeSpan.Zero)

        if initialized then
            notificationManager.Show(n)
        else
            pending.Enqueue(n)

    override _.OnKeyDown(e: Avalonia.Input.KeyEventArgs) =
        base.OnKeyDown(e)

        let isCtrlW =
            e.Key = Avalonia.Input.Key.W
            && e.KeyModifiers.HasFlag(Avalonia.Input.KeyModifiers.Control)

        if isCtrlW then this.Close()

    override _.OnClosing(e: WindowClosingEventArgs) =
        this.Hide()
        e.Cancel <- true
        base.Hide()

    override _.OnOpened(e) =
        base.OnOpened(e)

        if not initialized then
            initialized <- true

            while pending.Count > 0 do
                notificationManager.Show(pending.Dequeue())

File: src/TrayAppCS/Views/MainWindow.axaml.cs
using System;
using Avalonia.Controls;
using Avalonia.Controls.Notifications;
using Avalonia.Input;

namespace GlazeWM.TrayAppCS.Views;

public partial class MainWindow : Window
{
    private readonly WindowNotificationManager _notificationManager;
    private readonly System.Collections.Generic.Queue<Notification> _pending = new();
    private bool _initialized;

    public MainWindow()
    {
        InitializeComponent();
        _notificationManager = new WindowNotificationManager(this) { MaxItems = 3 };
    }

    public void Notify(string title, string body, NotificationType type = NotificationType.Information)
    {
        var notification = new Notification(title, body, type, TimeSpan.Zero);

        if (_initialized)
            _notificationManager.Show(notification);
        else
            _pending.Enqueue(notification);
    }

    protected override void OnKeyDown(KeyEventArgs e)
    {
        base.OnKeyDown(e);
        if (e.Key == Key.W && e.KeyModifiers.HasFlag(KeyModifiers.Control))
            Close();
    }

    protected override void OnClosing(WindowClosingEventArgs e)
    {
        Hide();
        e.Cancel = true;
        base.OnClosing(e);
    }

    protected override void OnOpened(EventArgs e)
    {
        base.OnOpened(e);

        if (!_initialized)
        {
            _initialized = true;
            while (_pending.TryDequeue(out var n))
                _notificationManager.Show(n);
        }
    }
}
